### PR TITLE
add wayland 1.18.0

### PR DESCRIPTION
**Pull request by ericLemanissier**
Specify library name and version:  **wayland/1.18.0**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by ericLemanissier** [comment]
that's surprising, the `.cmake` file seems to be in meson package:
```
meson/0.54.0: Appending PATH environment variable: /home/eric/.conan/data/meson/0.54.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin
[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] The *.cmake files have to be placed in a folder declared as `cpp_info.
builddirs`. Currently folders declared: [''] (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H019)
[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] Found files:
./bin/mesonbuild/cmake/data/preload.cmake (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H019)
```
How could https://github.com/conan-io/conan-center-index/pull/1215 pass ? @uilianries

**Comment by uilianries** [comment]
@ericLemanissier look the message again: "Found files: ./bin/mesonbuild/cmake/data/preload.cmake"

That cmake file must be removed. You could do:

```python
if self.options.shared and self.options.os == "Windows":
    tools.rmdir(os.path.join(self.package_folder, "bin", "mesonbuild"))
else:
    tools.rmdir(os.path.join(self.package_folder, "bin"))
```

File: recipes/libwayland/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def build(self):
+        meson = self._configure_meson()
+        meson.build()
+
+    def package(self):
+        self.copy(pattern="COPYING", dst="licenses", src=self._source_subfolder)
+        meson = self._configure_meson()
+        meson.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
```
Comment (uilianries) [inline recipes/libwayland/all/conanfile.py:74]:
```suggestion
        tools.rmdir(os.path.join(self.package_folder, "share"))
        tools.rmdir(os.path.join(self.package_folder, "bin"))
```

File: recipes/libwayland/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://wayland.freedesktop.org"
+    license = "MIT"
+    generators = "pkg_config"
+
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "enable_libraries": [True, False],
```
Comment (uilianries) [inline recipes/libwayland/all/conanfile.py:19]:
```suggestion
        "enable_libffi": [True, False],
```
I think more explicit sounds better

File: recipes/libwayland/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        "enable_dtd_validation": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "enable_libraries": True,
```
Comment (uilianries) [inline recipes/libwayland/all/conanfile.py:25]:
```suggestion
        "enable_libffi": True,
```

File: recipes/libwayland/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        "enable_dtd_validation": True,
+    }
+
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+    _meson = None
+    
+    def requirements(self):
+        if self.options.enable_libraries:
```
Comment (uilianries) [inline recipes/libwayland/all/conanfile.py:34]:
```suggestion
        if self.options.enable_libffi:
```

File: recipes/libwayland/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "wayland-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+        tools.replace_in_file(os.path.join(self._source_subfolder, 'meson.build'), "subdir('tests')", "#subdir('tests')")
+
+    def _configure_meson(self):
+        if not self._meson:
+            self._meson = Meson(self)
+            self._meson.configure(source_folder= self._source_subfolder, build_folder=self._build_subfolder, defs={
+                'libraries': 'true' if self.options.enable_libraries else 'false',
```
Comment (uilianries) [inline recipes/libwayland/all/conanfile.py:59]:
```suggestion
                'libraries': 'true' if self.options.enable_libffi else 'false',
```

**Comment by ericLemanissier** [comment]
Yes, I did check this `KB-H019` message, and it happens when applying the meson build_requirement, before getting wayland sources:
```
meson/0.54.0: Already installed!
meson/0.54.0: Appending PATH environment variable: /home/eric/.conan/data/meson/0.54.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/bin
[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] The *.cmake files have to be placed in a folder declared as `cpp_info.
builddirs`. Currently folders declared: [''] (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H019)
[HOOK - conan-center.py] post_package_info(): ERROR: [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] Found files:
./bin/mesonbuild/cmake/data/preload.cmake (https://github.com/conan-io/conan-center-index/wiki/Error-Knowledge-Base#KB-H019)
libwayland/1.18.0: Applying build-requirement: meson/0.54.0
libwayland/1.18.0: Applying build-requirement: ninja/1.9.0
[HOOK - conan-center.py] pre_source(): [IMMUTABLE SOURCES (KB-H010)] OK
libwayland/1.18.0: Configuring sources in /home/eric/.conan/data/libwayland/1.18.0/_/_/source
```

All the actual failures are happening for gcc 4.9 only because the error are the usual glibc ABI issue:
```
ninja: Entering directory `/tmp/c3ipr/pr_1325_2_0_9/.conan/data/libwayland/1.18.0/_/_/build/d8b83979afa10670feefec61a9bd3203127fe309/build_subfolder'
ninja: relocation error: ninja: symbol _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7reserveEm, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference
libwayland/1.18.0: 
```

File: recipes/libwayland/all/conanfile.py (hunk 1 above)
Comment (ericLemanissier) [reply recipes/libwayland/all/conanfile.py:74]: why would we remove bin ? it contains wayland-scanner

File: recipes/libwayland/all/conanfile.py (hunk 2 above)
Comment (ericLemanissier) [reply recipes/libwayland/all/conanfile.py:19]: this is actually about enabling [wayland libraries,](https://github.com/wayland-project/wayland/blob/1.18.0/meson_options.txt#L2) the fact that it requires libffi is an implementation detail

Comment (uilianries) [reply recipes/libwayland/all/conanfile.py:19]: So the package should be called wayland instead.

File: recipes/libwayland/all/conanfile.py (hunk 1 above)
Comment (uilianries) [reply recipes/libwayland/all/conanfile.py:74]: As the package is named LIBwayland, I thought the bin folder was irrelevant.

**Comment by uilianries** [comment]
@ericLemanissier I think with the new cross-building feature we can solve it. We can remove arch_build and os_build as a regular package.

**Comment by uilianries** [comment]
Or, just build using GCC 4.9

**Comment by ericLemanissier** [comment]
Regarding the name of the package, I'm not sure what should be done. I expect that conan will need to have a virtual package for system provided wayland, because I could not find information about the forward/backward compatibility of the wayland protocol. Some parts of the protocols are in their 7th version, and I don't know what happens if a wayland client (provided by conan)communicates with a wayland server (provided by the system package manager) with a different version.
The virtual package should be able to have a version, and raise an error if the package version is different from the version returned by `pkg-config --modversion wayland-client`
Another possibility would be to merge the virtual package into this one, by just adding a `use_system` option.

**Comment by Croydon** [comment]
> Another possibility would be to merge the virtual package into this one, by just adding a use_system option.

This could bring new confusion

This would look like `wayland/1.18.0:use_system=True`, so probably everyone would expect to use the version 1.18.0 and not just arbitrary version from the system.

And we can also not use something more obvious like `wayland/virtual` as this would conflict with `wayland/<version>`

**Comment by ericLemanissier** [comment]
> This would look like wayland/1.18.0:use_system=True, so probably everyone would expect to use the version 1.18.0 and not just arbitrary version from the system.
I agree that this is a bad situation, if the package does not check that the system installed version of the library is identical to the version of the conan package (using pkg-config)

**Comment by Croydon** [comment]
Random suggestion:

This "wayland" package [is described as](https://gitlab.freedesktop.org/wayland/wayland) `Core Wayland window system code and protocol`.

While wayland-protocols is described as `wayland-protocols contains Wayland protocols that add functionality not
available in the Wayland core protocol. Such protocols either add
completely new functionality, or extend the functionality of some other
protocol either in Wayland core, or some other protocol in
wayland-protocols.`

So if the name `libwayland` does not fit, maybe `wayland-core` is better?

By the way, both [Ubuntu](https://packages.ubuntu.com/search?keywords=libwayland&searchon=names&suite=eoan&section=all) and [Fedora](https://www.rpmfind.net/linux/rpm2html/search.php?query=wayland&submit=Search+...&system=&arch=) are splitting wayland into many small packages, but this might not be the best route for us? 

Ubuntu has a confusing `libwayland-bin` [package](https://packages.ubuntu.com/eoan/libwayland-bin)

**Comment by uilianries** [comment]
Pacman returns me:

```
extra/wayland 1.18.0-1 [installed]
    A computer display server protocol
...
multilib/lib32-wayland 1.18.0-1 [installed]
    A computer display server protocol
```

VCPKG and Conda do not offer wayland, it looks like we're the only crazy ones to try it haha

**Comment by ericLemanissier** [comment]
doesn't ` pkg-config --modversion wayland-client` return 1.18.0 ?
Regarding other package managers, I only found https://anaconda.org/pkgw-forge/wayland but I guess it's a community package ?

**Comment by uilianries** [comment]
@ericLemanissier if you want to add the binary, I recommend you renaming this recipe to wayland, as the original project name. The libwayland sounds like only the library.

**Pull request merged**